### ISSUE TITLE
more than 100000 partitions error occurred in 307

### ISSUE THREAD

**Issue by wyukawa** [prestosql/presto#619]
The following error occurs.
```
Query over table 'default.hoge' can potentially read more than 100000 partitions
```

```hoge``` table has two partition columns(service, yyyymmdd) and the number of total partition is more than 100000.
But the following query filters the partition columns and doesn't seem to scan more than 100000 partitions.
Actually, it didn't cause an error in prestosql 306 but the error occurred in 307.

query
```
select
 b.yyyymmdd
...
from (
 select
...
 from
  hoge
 where
  service in ('aaa)
 and
  yyyymmdd between '20171127' and '20180331'
 and
  ...
) a,
(
 select
  ...
 from
  hoge
 where
  service in ('bbb')
 and
  yyyymmdd between '20171127' and '20180331'
 and
  ...
 group by
  yyyymmdd
  ...
) b 
where 
 a.id = b.id 
 and
 a.yyyymmdd = b.yyyymmdd
group by
b.yyyymmdd
...
order by
b.yyyymmdd
...
```

**Comment by findepi** [comment]
@wyukawa is it possible to reproduce the problem when selecting one table at a time, eg.
```sql
 select
...
 from
  hoge
 where
  service in ('aaa)
 and
  yyyymmdd between '20171127' and '20180331'
```
?

cc @martint

**Comment by wyukawa** [comment]
@findepi Thank you for the quick response!
Unfortunately, I can't reproduce the problem when selecting one table.

Two table join is necessary to reproduce like this.
```
select
 b.yyyymmdd
from (
 select
  yyyymmdd
  ,json_extract_scalar(log, '$.id') as id
 from
  hoge
 where
  service in ('aaa')
 and
  yyyymmdd between '20171127' and '20180331'
) a,
(
 select
  yyyymmdd
  ,json_extract_scalar(log, '$.id') as id
 from
  hoge
 where
  service in ('bbb')
 and
  yyyymmdd between '20171127' and '20180331'
) b 
where 
 a.id = b.id 
 and
 a.yyyymmdd = b.yyyymmdd
limit 100
```

**Comment by ebyhr** [comment]
@findepi Perhaps, we can reproduce it by adding below code in TestHivePartitionsTable.java L149.
```java
        partitionListResult = query(format("SELECT a.* FROM (SELECT * FROM %s WHERE part_col = 1) a, (SELECT * FROM %s WHERE part_col = 1) b WHERE a.col = b.col", tableName, tableName));
        assertThat(partitionListResult).hasNoRows();
```

**Comment by ebyhr** [comment]
306 explain:
```
[[- Output[col, part_col] => [col:integer, part_col:integer]
        Estimates: {rows: 1 (10B), cpu: 138.00, memory: 14.00, network: 43.00}
    - RemoteExchange[GATHER] => [col:integer, part_col:integer]
            Estimates: {rows: 1 (10B), cpu: 138.00, memory: 14.00, network: 43.00}
        - InnerJoin[("col" = "col_7")][$hashvalue, $hashvalue_27] => [col:integer, part_col:integer]
                Estimates: {rows: 1 (10B), cpu: 138.00, memory: 14.00, network: 33.00}
                Distribution: PARTITIONED
            - RemoteExchange[REPARTITION][$hashvalue] => [col:integer, part_col:integer, $hashvalue:bigint]
                    Estimates: {rows: 1 (19B), cpu: 48.00, memory: 0.00, network: 19.00}
                - ScanProject[table = hive:HiveTableHandle{schemaName=default, tableName=tempto_mut_partitioned_table_with_variable_partitions_9rcyqfeg, analyzePartitionValues=Optional.empty}] => [col:integer, part_col:integer, $hashvalue_26:bigint]
                        Estimates: {rows: 1 (19B), cpu: 10.00, memory: 0.00, network: 0.00}/{rows: 1 (19B), cpu: 29.00, memory: 0.00, network: 0.00}
                        $hashvalue_26 := "combine_hash"(bigint '0', COALESCE("$operator$hash_code"("col"), 0))
                        col := col:int:0:REGULAR
                        part_col := part_col:int:-1:PARTITION_KEY
                            :: [[1]]
            - LocalExchange[HASH][$hashvalue_27] ("col_7") => [col_7:integer, $hashvalue_27:bigint]
                    Estimates: {rows: 1 (14B), cpu: 47.00, memory: 0.00, network: 14.00}
                - RemoteExchange[REPARTITION][$hashvalue_28] => [col_7:integer, $hashvalue_28:bigint]
                        Estimates: {rows: 1 (14B), cpu: 33.00, memory: 0.00, network: 14.00}
                    - ScanProject[table = hive:HiveTableHandle{schemaName=default, tableName=tempto_mut_partitioned_table_with_variable_partitions_9rcyqfeg, analyzePartitionValues=Optional.empty}] => [col_7:integer, $hashvalue_29:bigint]
                            Estimates: {rows: 1 (14B), cpu: 5.00, memory: 0.00, network: 0.00}/{rows: 1 (14B), cpu: 19.00, memory: 0.00, network: 0.00}
                            $hashvalue_29 := "combine_hash"(bigint '0', COALESCE("$operator$hash_code"("col_7"), 0))
                            col_7 := col:int:0:REGULAR
                            part_col:int:-1:PARTITION_KEY
                                :: [[1]]
]]
```

**Comment by sopel39** [comment]
Should it be considered a bug?

@ebyhr what is explain in 307 when you disable scan partition limit?

**Comment by ebyhr** [comment]
@sopel39 Here is 307 explain
```
[[- Output[col, part_col] => [col:integer, part_col:integer]
        Estimates: {rows: 1 (10B), cpu: 138.00, memory: 14.00, network: 43.00}
    - RemoteExchange[GATHER] => [col:integer, part_col:integer]
            Estimates: {rows: 1 (10B), cpu: 138.00, memory: 14.00, network: 43.00}
        - InnerJoin[("col" = "col_7")][$hashvalue, $hashvalue_27] => [col:integer, part_col:integer]
                Estimates: {rows: 1 (10B), cpu: 138.00, memory: 14.00, network: 33.00}
                Distribution: PARTITIONED
            - RemoteExchange[REPARTITION][$hashvalue] => [col:integer, part_col:integer, $hashvalue:bigint]
                    Estimates: {rows: 1 (19B), cpu: 48.00, memory: 0.00, network: 19.00}
                - ScanProject[table = hive:default:tempto_mut_partitioned_table_with_variable_partitions_te8ipsu8] => [col:integer, part_col:integer, $hashvalue_26:bigint]
                        Estimates: {rows: 1 (19B), cpu: 10.00, memory: 0.00, network: 0.00}/{rows: 1 (19B), cpu: 29.00, memory: 0.00, network: 0.00}
                        $hashvalue_26 := "combine_hash"(bigint '0', COALESCE("$operator$hash_code"("col"), 0))
                        col := col:int:0:REGULAR
                        part_col := part_col:int:-1:PARTITION_KEY
                            :: [[1]]
            - LocalExchange[HASH][$hashvalue_27] ("col_7") => [col_7:integer, $hashvalue_27:bigint]
                    Estimates: {rows: 1 (14B), cpu: 47.00, memory: 0.00, network: 14.00}
                - RemoteExchange[REPARTITION][$hashvalue_28] => [col_7:integer, $hashvalue_28:bigint]
                        Estimates: {rows: 1 (14B), cpu: 33.00, memory: 0.00, network: 14.00}
                    - ScanProject[table = hive:default:tempto_mut_partitioned_table_with_variable_partitions_te8ipsu8] => [col_7:integer, $hashvalue_29:bigint]
                            Estimates: {rows: 1 (14B), cpu: 5.00, memory: 0.00, network: 0.00}/{rows: 1 (14B), cpu: 19.00, memory: 0.00, network: 0.00}
                            $hashvalue_29 := "combine_hash"(bigint '0', COALESCE("$operator$hash_code"("col_7"), 0))
                            col_7 := col:int:0:REGULAR
                            part_col:int:-1:PARTITION_KEY
                                :: [[1]]
]]
```

and diff
```diff
10c10
<                 - ScanProject[table = hive:HiveTableHandle{schemaName=default, tableName=tempto_mut_partitioned_table_with_variable_partitions_9rcyqfeg, analyzePartitionValues=Optional.empty}] => [col:integer, part_col:integer, $hashvalue_26:bigint]
---
>                 - ScanProject[table = hive:default:tempto_mut_partitioned_table_with_variable_partitions_te8ipsu8] => [col:integer, part_col:integer, $hashvalue_26:bigint]
20c20
<                     - ScanProject[table = hive:HiveTableHandle{schemaName=default, tableName=tempto_mut_partitioned_table_with_variable_partitions_9rcyqfeg, analyzePartitionValues=Optional.empty}] => [col_7:integer, $hashvalue_29:bigint]
---
>                     - ScanProject[table = hive:default:tempto_mut_partitioned_table_with_variable_partitions_te8ipsu8] => [col_7:integer, $hashvalue_29:bigint]
```

**Comment by ngpgopi** [comment]
@findepi I am facing the same problem even on single table. 

select * from data limit 100;
> Query 20190411_164620_00171_vq9g8 failed: Query over table 'sdii_parquet.data' can potentially read more than 100000 partitions

**Comment by ngpgopi** [comment]
@findepi I am facing the same problem even on single table. 

select * from data limit 100;
> Query 20190411_164620_00171_vq9g8 failed: Query over table 'sdii_parquet.data' can potentially read more than 100000 partitions

**Comment by ngpgopi** [comment]
@findepi I am facing the same problem even on single table. 

select * from data limit 100;
> Query 20190411_164620_00171_vq9g8 failed: Query over table 'sdii_parquet.data' can potentially read more than 100000 partitions

**Comment by martint** [comment]
@ngpgopi, do you see that issue in earlier versions? Can you post the full stacktrace?

**Comment by findepi** [comment]
@ngpgopi for query like `select * from data limit 100;` we would always be hitting "can potentially read more than 100000 partitions".
When listing partitions we unfortunately do not know whether they are empty or not, so we don't know whether listing 100 partitions (or 1,000, or 10,000) is enough.
I appreciate the fact that in most of the cases it _is_ enough to get first 100 partitions (usually: only 1 partition), but we need to be 100% correct here.

As a workaround, you please add some condition on partitioning keys.
If you do not know the partitioning keys, you can always run:
```
select * from "data$partitions" -- with "limit 100" or without
```
This query should work even if the table has more than 100,000 partitions.

**Comment by martint** [comment]
Fixed by #621

**Comment by electrum** [comment]
I added the partition limit two years ago in order to fix a production issue where queries on tables with millions of partitions would cause the coordinator to do a full GC due to memory pressure. It is a safety feature, but you can experiment with increasing or (effectively) removing the limit by changing the `hive.max-partitions-per-scan` Hive config property.

There are probably changes we could make to the Hive connector to support this use case, but it would likely involve increased complexity and other tradeoffs. Even at Facebook, tables above this partition limit were rare, and we considered such tables to be "abusing" partitioning, so making this work was not a priority.
